### ISSUE TITLE
Implement learned answer quality scorer

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#547]
## Overview

Implement **learned quality scorer** to validate and rank answers from all extraction methods.

This provides confidence scores for answers regardless of extraction method used.

**Part of Epic**: #543  
**Depends on**: #544 (AST extraction)  
**Priority**: MEDIUM (needed for ranking multiple candidates)

### Problem

Different extraction methods return answers with varying quality:
- AST: High precision, but may miss answers
- Heuristic: Lower precision, more false positives
- Learned: Variable quality

**Need**: Unified quality score to rank/filter answers

### Solution

Train a model to predict: **"Is this answer correct for this query?"**

### Model Architecture

**Lightweight binary classifier**:

```
Input: (Query, Sentence, Candidate Answer)
  ↓
Encode with CompositionalEmbedding
  ↓
Features:
  - Query-answer semantic similarity
  - Answer type match (person for WHO, location for WHERE)
  - Position in sentence (subject > object > modifier)
  - Extraction method confidence
  ↓
Binary classifier: Correct (1) or Incorrect (0)
  ↓
Output: Quality score [0-1]
```

## Implementation

**File**: `klareco/models/answer_quality_scorer.py`

```python
import torch
import torch.nn as nn

class AnswerQualityScorer(nn.Module):
    """
    Learned model to score answer quality.
    
    Predicts probability that candidate answer is correct.
    """
    
    def __init__(self, comp_emb, hidden_dim=128):
        super().__init__()
        self.comp_emb = comp_emb
        embed_dim = comp_emb.embed_dim
        
        # Encoders
        self.query_encoder = nn.LSTM(embed_dim, hidden_dim, batch_first=True)
        self.answer_encoder = nn.LSTM(embed_dim, hidden_dim, batch_first=True)
        
        # Feature combination
        self.scorer = nn.Sequential(
            nn.Linear(hidden_dim * 2 + 10, hidden_dim),  # +10 for hand-crafted features
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(hidden_dim, 1),
            nn.Sigmoid()
        )
    
    def forward(self, query_words, answer_words, features):
        """
        Args:
            query_words: List of word structures for query
            answer_words: List of word structures for answer
            features: Tensor [10] - hand-crafted features
        
        Returns:
            quality_score: Float [0-1]
        """
        # Encode query and answer
        query_embs = self._encode_words(query_words)
        answer_embs = self._encode_words(answer_words)
        
        # LSTM encode
        _, (query_hidden, _) = self.query_encoder(query_embs.unsqueeze(0))
        _, (answer_hidden, _) = self.answer_encoder(answer_embs.unsqueeze(0))
        
        # Concatenate
        query_vec = query_hidden.squeeze(0)  # [hidden_dim]
        answer_vec = answer_hidden.squeeze(0)  # [hidden_dim]
        
        combined = torch.cat([query_vec, answer_vec, features])  # [hidden*2 + 10]
        
        # Score
        quality = self.scorer(combined)
        
        return quality

def extract_features(query_ast, answer_ast, extraction_method, position):
    """Extract hand-crafted features for quality scoring."""
    features = []
    
    # 1. Question-answer type match
    features.append(question_type_matches_answer(query_ast, answer_ast))
    
    # 2. Answer position in sentence (subject=1.0, object=0.7, modifier=0.5)
    features.append(position_score(position))
    
    # 3. Extraction method confidence
    features.append(extraction_method.confidence)
    
    # 4. Answer length (words) - normalized
    features.append(len(answer_words) / 10.0)
    
    # 5. Capitalization (proper noun indicator)
    features.append(is_capitalized(answer_text))
    
    # ... more features ...
    
    return torch.tensor(features)
```

### Positive Examples

From sentences where AST extraction succeeds with high confidence:

```python
sentence = "Zamenhof fondis Esperanton en 1887."
query = "Kiu fondis Esperanton?"
answer = "Zamenhof"  # Correct!

training_example = {
    'query': query,
    'sentence': sentence,
    'answer': answer,
    'label': 1.0  # Correct
}
```

### Negative Examples

Generate incorrect answers:
1. **Wrong entity**: Extract different noun from sentence
2. **Type mismatch**: Year for WHO question, person for WHEN
3. **Random text**: Arbitrary span that's not an answer

```python
# Negative example 1: Wrong entity
{
    'query': "Kiu fondis Esperanton?",
    'sentence': "Zamenhof fondis Esperanton en 1887.",
    'answer': "1887",  # Wrong! (this is WHEN, not WHO)
    'label': 0.0
}

# Negative example 2: Type mismatch
{
    'query': "Kiam fondis Zamenhof Esperanton?",
    'sentence': "Zamenhof fondis Esperanton en 1887.",
    'answer': "Zamenhof",  # Wrong! (this is WHO, not WHEN)
    'label': 0.0
}
```

**Target**: 20,000+ examples (10k positive, 10k negative)

## Success Criteria

- ✅ Achieves 80%+ accuracy on validation set
- ✅ Calibrated confidence (predicted score ≈ actual accuracy)
- ✅ Improves ranking: Correct answers scored higher than incorrect
- ✅ Fast inference (<50ms per answer)

## Training Script

**Script**: `scripts/train_answer_quality_scorer.py`

```bash
# Generate training data
python scripts/generate_answer_quality_data.py \
  --corpus data/enhanced_corpus/corpus_with_metadata.jsonl \
  --output-dir data/training/answer_quality/

# Train model
python scripts/train_answer_quality_scorer.py \
  --train-data data/training/answer_quality/train.jsonl \
  --val-data data/training/answer_quality/val.jsonl \
  --comp-model models/root_embeddings/best_model.pt \
  --output-dir models/answer_quality_scorer/ \
  --epochs 20 \
  --batch-size 64
```

### Usage in Answer Extractor

```python
class AnswerExtractor:
    def __init__(self, quality_scorer):
        self.ast_extractor = ASTAnswerExtractor()
        self.heuristic_extractor = HeuristicAnswerExtractor()
        self.quality_scorer = quality_scorer
    
    def extract(self, query, sentence):
        candidates = []
        
        # Try AST extraction
        ast_answer = self.ast_extractor.extract(query, sentence)
        if ast_answer:
            candidates.append(ast_answer)
        
        # Try heuristic extraction
        heur_answer = self.heuristic_extractor.extract(query, sentence)
        if heur_answer:
            candidates.append(heur_answer)
        
        # Score all candidates with quality scorer
        for candidate in candidates:
            quality_score = self.quality_scorer(query, sentence, candidate['answer'])
            candidate['quality_score'] = quality_score
        
        # Return best candidate (highest quality score)
        candidates.sort(key=lambda x: x['quality_score'], reverse=True)
        return candidates[0] if candidates else None
```

### Semantic Features (from embeddings)
1. Query-answer similarity (cosine)
2. Query-sentence similarity
3. Answer-sentence relevance

### Syntactic Features (from AST)
4. Answer position (subject/object/modifier)
5. Answer role matches question type
6. Distance from query verb

### Surface Features
7. Answer length (words)
8. Capitalization pattern
9. Contains numbers/dates
10. Extraction method + confidence

## Timeline

**Estimated**: 2 days
- Morning Day 1: Generate training data
- Afternoon Day 1: Implement model
- Day 2: Train, evaluate, integrate

## Dependencies

- Requires: #544 (AST extractor) to generate positive examples
- Requires: CompositionalEmbedding (exists)
- Optional: #545 (Heuristic) for more diverse training data

## Related Issues

- Part of Epic: #543
- Improves: #544, #545 (scores their outputs)
- Independent of: #546 (can implement without learned extraction)

## Metrics to Track

**During Training**:
- Binary accuracy (correct/incorrect)
- AUC-ROC (ranking quality)
- Calibration (predicted score vs actual accuracy)

**After Integration**:
- Impact on end-to-end QA accuracy
- Correlation with human judgments
- False positive rate (incorrect answers scored high)

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
